### PR TITLE
Angular12

### PR DESCRIPTION
**Pull request by astone2014**
# Description

Update to Angular 12

## Issues Resolved

https://github.com/johannesjo/angular-material-css-vars/issues/67

## Check List

- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable.

### CODE REVIEW

**Comment by astone2014** [comment]
This is just a start. There is more work to do.

**Comment by johannesjo** [comment]
Thank you very much @astone2014 ! I'll have a look tomorrow!

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "angular-material-css-vars",
-  "version": "1.2.0",
+  "version": "1.2.1",
```
Comment (johannesjo) [inline package.json:3]: This is something for later :)

**Comment by johannesjo** [comment]
This is a very promising start! Please let me know, if I can support you somehow moving forward.

**Comment by astone2014** [comment]
I just need to find time.

There are plenty of problems still. I really only hacked together the css variables with the basic palette.

Still learning your library will keep you updated.

**Comment by astone2014** [comment]
I opened an issue with the material components library about dealing with opacity.
https://github.com/angular/components/issues/22759

I'm not sure how we can override the sass function exported by `@use`.

**Comment by pedrojrivera** [comment]
For what it's worth, our applications here rely heavily on variables for its theming. So when we upgraded to v12, theming broke and nothing was working.
 
We use a modified version of material-css-vars in our projects, and we had to make some modifications in order to get it to work for us. Many of the changes we made are very similar to what @astone2014  made with some additional modifications after we did some debugging. We now have a project that is fully functional. 

I will fork this project and apply the same changes we made for your review. 

Thanks!

**Comment by pedrojrivera** [comment]
I created a branch of the changes we have made so far. I made sure to preserve all the existing functionality while trying to work around the Angular namespace issue.

Unless the Angular team provides some sort of callback functionality in the get-color-from-palette() when it detects a variable, you won't be able to control the css for the angular internals that use theme.get-color-from-palette(). From our testing, in most cases this is not an issue as we can target the css and add additional overwrites.

We had to add additional overwrites to mat-slide-toggle and mat-checkbox.  

https://github.com/pedrojrivera/angular-material-css-vars/commits/feature/angular-vars-v12

Thanks!

**Comment by johannesjo** [comment]
Thank you very much  @pedrojrivera ! 

> Unless the Angular team provides some sort of callback functionality in the get-color-from-palette() when it detects a variable, you won't be able to control the css for the angular internals that use theme.get-color-from-palette(). From our testing, in most cases this is not an issue as we can target the css and add additional overwrites.

For me, this is good enough for now. For people who need this, there is always the option to use version 11 of angular material.

**Comment by pedrojrivera** [comment]
@johannesjo I totally agree. We will most likely stay on v11 unless Angular adds some new functionality to make theming easier. It usually takes us 2-3 months to fully migrate to the next version due to dependencies. I believe the Angular team will be fully implementing MDC which is supposed to enable the use of variables. Maybe by v13. ??

**Comment by pedrojrivera** [comment]
We made some additional changes to stabilize a few things. Some changes we made were simply to lean out some of the variables to get a better understanding of what is going on. In doing that we fixed:

1. Fixed duplicating themes error message
2. Removed excess variable output at the root.
3. Removed light/dark text nested variable use.

**Comment by astone2014** [comment]
Nice job @pedrojrivera 

I reset my branch to yours.

I reviewed your code and am happy with the changes.

**Comment by pedrojrivera** [comment]
@astone2014 In the get-color-from-palette(), there is a block of code that no longer has any meaning.  Lines 14 - 20:

```
$inner-var-alpha-str: str-slice($color, 5, -2) + '-alpha';

// if we have a corresponding alpha variable use it
// don't use alpha variable if we have an actual $opacity value
@if (map_get($mat-css-default-light-theme, $inner-var-alpha-str) and $opacity == null) {
  @return rgba($color, _mat-css-str-replace($color, ')', "-alpha)"));
}
```

In the _variables.scss, the foreground and background variables for light and dark theme were removed and reset to use material's internal maps. 

The reasoning for this that since we truly can't override the get-color-from-palette(), we need to rely on variable structures that are friendly to Material's internal processing.

**Comment by pedrojrivera** [comment]
After some more testing we discovered a few more issues. I added a new commit to fix an issue where alpha values were not applied correctly when using variables.

https://github.com/pedrojrivera/angular-material-css-vars/commit/32984c0fea6687bbcbb63b6c631e7c2dbcc0c7bf

**Comment by nekkon** [comment]
Would be great to have this merged, if rdy. It is the only reason left not to upgrade to Angular12.

**Comment by johannesjo** [comment]
What is the state of things @astone2014 ?

**Comment by futbotism** [comment]
@johannesjo +1 with getting this merged. Is there any tasks that I can undertake to help along with this PR?

**Comment by johannesjo** [comment]
@futbotism not sure what is up with @astone2014 . But maybe you can open up a new PR with the changes and we continue to work on it there?

**Comment by futbotism** [comment]
I could, but im not entirely sure which changes are required

**Comment by futbotism** [comment]
@johannesjo Are there any tasks that I can commit to this PR?

**Comment by johannesjo** [comment]
@futbotism you can follow the code provided in this PR and in the frok by @pedrojrivera :
https://github.com/pedrojrivera/angular-material-css-vars/commit/32984c0fea6687bbcbb63b6c631e7c2dbcc0c7bf

**Comment by pedrojrivera** [comment]
We upgraded one of our projects using the latest code from my fork. So far we have had no issues. We have tested it pretty thoroughly and we are pleased with the results.  

I would recommend, once upgraded, just do a thorough test.

**Comment by pedrojrivera** [comment]
I created a new PR: https://github.com/johannesjo/angular-material-css-vars/pull/72 with my latest commits.

Please note:  We have a very large project that we have tested with and so far so good, no issues. With that said, we don't use every Material component, so as others test it is possible that some additional overrides may be necessary. Thanks!

**Comment by johannesjo** [comment]
I am closing this in favor of: 
https://github.com/johannesjo/angular-material-css-vars/pull/72

**Pull request closed without merging**